Add debug backend that applies CrossRefFakeMode, use in compiler bisector

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138651


I was debugging an internal ne divergence for a while that ended up being because of a bad meta. I added an explicit a config option and an explicit backend `aot_eager_decomp_partition_crossref` to enable the FakeCrossRefMode when running the graph.  I added an explicit backend bc I suspect it will be useful for internal models but I'm also happy to leave as config option.

It will only test ops that have meta to avoid memory overhead of hitting fallback path and running in eager.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

eellison committed:
Add debug backend that applies CrossRefFakeMode, use in compiler bisector

[ghstack-poisoned]

zou3519 reviewed (approved):
I don't have thoughts on backend vs config, maybe Brian does. It seems easy enough to switch or to add both, so approving.

zou3519 commented on torch/_subclasses/fake_utils.py:
```diff
@@ -82,18 +82,21 @@ def __init__(
         kwargs = kwargs or {}
 
         fake_r = None
+        breakpoint()
```
don't think this is supposed to be here? :P

zou3519 commented on torch/_functorch/config.py:
```diff
@@ -162,6 +162,10 @@ def remote_autograd_cache_default() -> Optional[bool]:
 
+
+# Run aot eager decomp partition with CrossRefFakeMode
+fake_tensor_crossref = False
```
to check, your PR made this both a config option and a backend?

eellison replied:
that's correct

eellison committed:
Update on "Add debug backend that applies CrossRefFakeMode, use in compiler bisector"



I was debugging an internal ne divergence for a while that ended up being because of a bad meta. I added an explicit a config option and an explicit backend `aot_eager_decomp_partition_crossref` to enable the FakeCrossRefMode when running the graph.  I added an explicit backend bc I suspect it will be useful for internal models but I'm also happy to leave as config option.

It will only test ops that have meta to avoid memory overhead of hitting fallback path and running in eager.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov rec

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

bdhirsh reviewed (approved):
sgtm, agreed that piping configs through internal can be a pain and an extra backend seems harmless

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge -f

eellison commented:
@pytorchbot merge -f "merge taking forever"
